crates/autopilot/src/infra/persistence/dto/quote.rs

```diff
@@ -1,25 +1,46 @@
 use {
     crate::{boundary, domain},
-    number::conversions::big_decimal_to_u256,
-    primitive_types::U256,
+    bigdecimal::{
+        num_traits::{CheckedDiv, CheckedMul},
+        FromPrimitive,
+    },
+    num::BigRational,
+    number::conversions::{big_decimal_to_u256, big_rational_to_u256},
 };
 
-pub fn into_domain(
-    quote: boundary::database::orders::Quote,
-) -> Result<domain::Quote, AmountOverflow> {
+pub fn into_domain(quote: boundary::database::orders::Quote) -> Result<domain::Quote, QuoteError> {
+    let gas_amount = BigRational::from_f64(quote.gas_amount).ok_or(QuoteError::InvalidInput)?;
+    let gas_price = BigRational::from_f64(quote.gas_price).ok_or(QuoteError::InvalidInput)?;
+    let sell_token_price =
+        BigRational::from_f64(quote.sell_token_price).ok_or(QuoteError::InvalidInput)?;
+    let fee = big_rational_to_u256(
+        &gas_amount
+            .checked_mul(&gas_price)
+            .ok_or(QuoteError::AmountOverflow)?
+            .checked_div(&sell_token_price)
+            .ok_or(QuoteError::DivisionByZero)?,
+    )
+    .map_err(QuoteError::Error)?;
     Ok(domain::Quote {
         order_uid: domain::OrderUid(quote.order_uid.0),
         sell_amount: big_decimal_to_u256(&quote.sell_amount)
-            .ok_or(AmountOverflow)?
+            .ok_or(QuoteError::AmountOverflow)?
             .into(),
         buy_amount: big_decimal_to_u256(&quote.buy_amount)
-            .ok_or(AmountOverflow)?
-            .into(),
-        fee: U256::from_f64_lossy(quote.gas_amount * quote.gas_price / quote.sell_token_price)
+            .ok_or(QuoteError::AmountOverflow)?
             .into(),
+        fee: fee.into(),
     })
 }
 
 #[derive(Debug, thiserror::Error)]
-#[error("amount overflow")]
-pub struct AmountOverflow;
+pub enum QuoteError {
+    #[error("amount overflow")]
+    AmountOverflow,
+    #[error("invalid input")]
```

Comment:
Can you be more precise, mentioning that it cannot be converted to a valid rational number

Reply:
Done